### PR TITLE
[1505] Add ci script and integration with dev container, local dev env, pipeline & pre-commit hook. 

### PR DESCRIPTION
**Pull request by argoldbe**
- Add ci.sh and setup_ci.sh scripts to be used in dev container, local dev env and in pipelines
- Updates to dev container config to support ci scripts
- Add pre-commit integration & config
- Add detect-secrets integration
- Updates to CONTRIBUTION.md to include CI details (this will be enhanced in future as part of 'Development on Windows' section updates)

### CODE REVIEW

File: RetrospectiveExtension.Frontend/.markdownlint.json
Comment (argoldbe) [inline RetrospectiveExtension.Frontend/.markdownlint.json:2]: Disabling the default markdown lint rules to limit PR to CI related work. Will enable in future PR with required markdown changes.

File: .secrets.baseline
Comment (mhassaninmsft) [inline .secrets.baseline:1]: How was this file generated?

File: .devcontainer/devcontainer.json
Comment (mhassaninmsft) [inline .devcontainer/devcontainer.json:27]: This extension which is powered by `markdownlint` node package, the same engine powering `markdownlint-cli` is laready installed in the dev container and enabled by Engin. I was wondering why markdown linting warning errors were showing for me. :)

File: RetrospectiveExtension.Frontend/.markdownlint.json
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/.markdownlint.json:2]: Why is the `.markdownlint` located inside `RetrospectiveExtension.Frontend` The Markdown files are at the top level of the directory. Should not this file be at the top level?

Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/.markdownlint.json:2]: nvm, I see that it is only applicable to the assets directory

Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/.markdownlint.json:2]: Interestingly all the violations are max length violations, and they can't be quick fixed

File: .secrets.baseline
Comment (argoldbe) [reply .secrets.baseline:1]: @mhassaninmsft  The file was generated by the detect-secrets module. I executed 'detect-secrets scan > .secrets.baseline'. An initial baseline file is required when running the tool. Once this PR is approved my plan is to rebase and generate/commit an updated baseline file prior to completing the PR.

File: .github/workflows/ci_frontend.yml
Comment (phongcao) [inline .github/workflows/ci_frontend.yml:38]: I think that we don't need this line since bash is the default shell (same as other run steps in the file).

File: .github/workflows/ci_frontend.yml
Comment (phongcao) [inline .github/workflows/ci_frontend.yml]: Should we change the name to something clearer? Such as `Build and Test` like the old step.

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (phongcao) [inline RetrospectiveExtension.Frontend/scripts/ci.sh:20]: Did we forget adding the `npm run pack:p` command? This helps validate the extension json file.

**Review by phongcao — CHANGES_REQUESTED** [review]

File: RetrospectiveExtension.Frontend/scripts/setup_ci.sh
Comment (phongcao) [inline RetrospectiveExtension.Frontend/scripts/setup_ci.sh:8]: Considering [Task #1499](https://dev.azure.com/csebostoncrew/Retrospectives-AzDO-Extension/_workitems/edit/1499), should we worry about locking npm version as well?

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/scripts/ci.sh:15]: One concern I have is that the `ci.sh` would fail locally since there is no `enviornment.tsx` file when you first clone the repo and start working. To use the pre-commit hook successfully you have to create that file. This is ok most of the times, but for cases where somebody clones the repo to update some Markdown lines or add some extra comments somewhere, the pre-commit would fail which is not a great experience. I would suggest checking if the `/config/enviornemnt.tsx` file is not found, then creating it if it is not found,

File: RetrospectiveExtension.Frontend/scripts/setup_ci.sh
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/scripts/setup_ci.sh:8]: There will be only one npm and node installed in the devContainer, and that is the npm in the `DockerFile` which will also be used here. What do you have in mind here?

**Review by mhassaninmsft — CHANGES_REQUESTED** [review]
This is great! Will sure make our lives easier.
 I only have 1 concern related to the `enviornment.tsx` on fresh local repos as well as Phong's comment on `pack:p`

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:20]: I can add that back. I was focusing on CI but you make a good point about validating the manifest.

File: RetrospectiveExtension.Frontend/.markdownlint.json
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/.markdownlint.json:2]: Yes - will enable and fix in future PR.

File: RetrospectiveExtension.Frontend/scripts/setup_ci.sh
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/scripts/setup_ci.sh:8]: @phongcao @mhassaninmsft Let's hold this for a future PR - we can discuss at the next parking lot.

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:15]: @mhassaninmsft That's a good point. Since this file is required to build I'm wondering if we should remove the template and just check in the environment.tsx with the placeholder values to simplify. I don't think there is any downside from a dev/test perspective if the backend is not being exercised.

File: RetrospectiveExtension.Frontend/scripts/setup_ci.sh
Comment (phongcao) [reply RetrospectiveExtension.Frontend/scripts/setup_ci.sh:8]: Agreed

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:20]: @phongcao  I ran into an interesting problem when adding pack:p. Since the pre-commit operation stashes unstaged changes while running hooks the update of the manifest (version increment) is conflicting with the stash apply and the commit is rolled back. The error message is "[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...". To work around this I added a new pack:test script that does not include the --rev-version flag. This will validate the manifest and allow the pre-commit hook to operate as intended.

Comment (phongcao) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:20]: Make senses, thanks for explaining.

File: RetrospectiveExtension.Frontend/scripts/setup_ci.sh
Comment (phongcao) [inline RetrospectiveExtension.Frontend/scripts/setup_ci.sh]: nit: looks like the indentation here is different with the other script.

**Review by phongcao — APPROVED** [review]

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:15]: It would make the development expirence worse I think, this was the case before creating the environment template. Many people including me would accedientlay check in the application inssghts instrumentation key and the url for the backend. Ideally for config file they are in Json files or loaded dynamically from the env. Here the file is actually a jsx file which is required at build time. The best solution would be to move `enviorment.tsx` into `environment` and require it at run time. I think Json files should not cause a build failure at compile time. I will verify that. Anyway I leave that final decision up to you.

**Review by mhassaninmsft — APPROVED** [review]

File: RetrospectiveExtension.Frontend/scripts/ci.sh
Comment (argoldbe) [reply RetrospectiveExtension.Frontend/scripts/ci.sh:15]: Plan is to address this in a future PR by switching to a proper config mechanism. Leaving as is for now with the known requirement that the `environment.tsx` file is copied/created prior to performing a commit.

**Pull request merged**
